SDWebImage/SDWebImageCodersManager.m

```diff
@@ -110,7 +116,10 @@ - (UIImage *)decompressedImageWithImage:(UIImage *)image
     if (!image) {
         return nil;
     }
-    for (id<SDWebImageCoder> coder in self.coders) {
+    LOCK(self.codersLock);
+    NSArray<id<SDWebImageCoder>> *coders = self.coders;
+    UNLOCK(self.codersLock);
+    for (id<SDWebImageCoder> coder in coders.reverseObjectEnumerator) {
```

Comment:
this looks a bit weird. Usually, you would lock the enumeration as well. Here it works because we actually can't alter `self.coders`, we can only change the pointer to another array.

Reply:
Actually :)...If the coders during enumeration is modified by another thread, there is no way to protect it. Because

```objective-c
Note

In Objective-C, it is not safe to modify a mutable collection while enumerating through it. Some enumerators may currently allow enumeration of a collection that is modified, but this behavior is not guaranteed to be supported in the future.
```

So, the best choice, it's to keep a `copy` of current state. Which not get effected by later add or removal.
The best choice, since `self.coders` is a immutable `NSArray`, just reference it is OK.

Reply:
Swift's Array is a struct. But however, it using `copy-on-write`, which is not compatible for thread-safe. So you should also take attention the for-in loop during different threads are not safe. Create a copy of current state is better.